### PR TITLE
ptx: Improve performance: remove N^2 loop and reduce repetitive regex compilations.

### LINKED ISSUE #1588: ptx hangs on large inputs in roff mode due to suboptimal algorithm
[The recent study](https://arxiv.org/pdf/2008.06537.pdf) have fuzzed the hell out of coreutils. As collateral, they discovered that `ptx` effectively hangs on large inputs due to suboptimal processing algorithm:

> In Rust, ptx appeared to hang because of a N2 algorithm in
function format_roff_line. This function processes almost
every substring on the input line that ends with the last char-
acter. Since our input was a single line of 93,290,954 bytes, N2
is 8.7 × 1015. We ran this test case for 34 hours and it ﬁnished
22% of the input, so we estimate completion time in 155 hours.
While technically not a hang, this ran long enough that we
considered it one.

I also noticed that this function recompiles the same regex 4-5 times on each call. Unnecessary allocations are aplenty as well.

As a bit of realistic point of view, `ptx` is used by approximately nobody these days, so I'm not really sure if spending time on improving the program worth anything.

### PR DESCRIPTION
**Pull request by wishawa**
Fixes #1588 

## Benchmarks:
Release build on Linux on a quad core Intel CPU.

File: [http://textfiles.com/stories/aesop11.txt](http://textfiles.com/stories/aesop11.txt) (220 kB).
(empty lines were removed because GNU ptx -G couldn't handle them)
#### uutils ptx before my patch
```
user@pc:~/Rust-Projects/uutils$ time ./uu-ptx-old -G long_text_with_newlines.txt > /dev/null

real	0m1.335s
user	0m1.323s
sys	0m0.008s
```
#### uutils ptx with my patch
```
user@pc:~/Rust-Projects/uutils$ time ./uu-ptx-new -G long_text_with_newlines.txt > /dev/null

real	0m0.083s
user	0m0.077s
sys	0m0.006s
```
#### GNU ptx on Fedora 32
```
user@pc:~/Rust-Projects/uutils$ time ptx -G long_text_with_newlines.txt > /dev/null

real	0m0.038s
user	0m0.034s
sys	0m0.004s
```
#### uutils ptx before my patch
```
user@pc:~/Rust-Projects/uutils$ time ./uu-ptx-old -G long_text_without_newlines.txt > /dev/null

real	0m28.478s
user	0m28.325s
sys	0m0.008s
```
#### uutils ptx with my patch
```
user@pc:~/Rust-Projects/uutils$ time ./uu-ptx-new -G long_text_without_newlines.txt > /dev/null

real	0m0.111s
user	0m0.099s
sys	0m0.010s
```
#### GNU ptx on Fedora 32
```
user@pc:~/Rust-Projects/uutils$ time ptx -G long_text_without_newlines.txt > /dev/null

real	0m0.041s
user	0m0.037s
sys	0m0.003s
```

### CODE REVIEW

**Comment by wishawa** [comment]
I'm still a beginner in Rust, so if I'm doing anything wrong or unidiomatic, please do tell!

**Comment by rivy** [comment]
@wishawa , thanks for the contribution!

The story is a bit hard to follow here ... This looks like only three changes to the original code base, so to help me review this, please squash the "fixes" for your additions into the original commit. For my initial investigation, it looks like the first five should likely be a single commit. And then that commit could be separated into the regex change and the loop change.

As for actual code critique, it appears that `WS_REG` (and the associated `lazy_static` package) aren't needed. The relevant regex (`config.context_regex`) that could be used to separate content within the line is already included in the configuration information passed to the `read_input` function.

If you'll make those changes, I'll look back at this more closely.

Thanks again.

**Comment by wishawa** [comment]
Sorry for the messy commits 😅. I've force pushed the reorganized version over.

For WS_REG, my understanding is it is not the same thing as context_regex. I believe it is used to replace whitespace characters with plain space " ". But is_whitespace() does that anyway so I say we just remove it.

(There is probably some small performance gain to be had by replacing all whitespaces with plain space and only comparing against it, but let's do that later)

File: src/uu/ptx/src/ptx.rs [hunk 1]
```diff
@@ -203,25 +211,29 @@ fn read_input(input_files: &[String], config: &Config) -> HashMap<String, (Vec<S
             Box::new(file)
         });
         let lines: Vec<String> = reader.lines().map(|x| crash_if_err!(1, x)).collect();
+        let chars_lines: Vec<Vec<char>> = lines.iter().map(|x| x.chars().collect()).collect();
```
Comment (sylvestre) [inline src/uu/ptx/src/ptx.rs:214]: please add a comment explaining what it is doing

File: src/uu/ptx/src/ptx.rs [hunk 2]
```diff
@@ -203,25 +211,29 @@ fn read_input(input_files: &[String], config: &Config) -> HashMap<String, (Vec<S
...
     }
     file_map
 }
 
-fn create_word_set(
-    config: &Config,
-    filter: &WordFilter,
-    file_map: &HashMap<String, (Vec<String>, usize)>,
-) -> BTreeSet<WordRef> {
+fn create_word_set(config: &Config, filter: &WordFilter, file_map: &FileMap) -> BTreeSet<WordRef> {
```
Comment (sylvestre) [inline src/uu/ptx/src/ptx.rs:229]: please document this function

File: src/uu/ptx/src/ptx.rs [hunk 3]
```diff
@@ -316,57 +327,73 @@ fn trim_idx(s: &[char], beg: usize, end: usize) -> (usize, usize) {
...
-    let mut head = String::new();
-    let mut before = String::new();
-    let mut after = String::new();
-    let mut tail = String::new();
-
-    let half_line_size = cmp::max(
-        (config.line_width / 2) as isize - (2 * config.trunc_str.len()) as isize,
+    let half_line_size = (config.line_width / 2) as usize;
+    let max_before_size = cmp::max(half_line_size as isize - config.gap_size as isize, 0) as usize;
+    let max_after_size = cmp::max(
```
Comment (sylvestre) [inline src/uu/ptx/src/ptx.rs:337]: please add a comment explaining that it is doing

File: src/uu/ptx/src/ptx.rs [hunk 4]
```diff
@@ -316,57 +327,73 @@ fn trim_idx(s: &[char], beg: usize, end: usize) -> (usize, usize) {
...
+    let mut after = String::with_capacity(half_line_size);
+    let mut tail = String::with_capacity(half_line_size);
 
     // get before
-    let mut bb_tmp = cmp::max(all_before.len() as isize - max_before_size as isize, 0) as usize;
-    bb_tmp = trim_broken_word_left(&all_before_vec, bb_tmp, all_before.len());
-    let (before_beg, before_end) = trim_idx(&all_before_vec, bb_tmp, all_before.len());
-    before.push_str(&all_before[before_beg..before_end]);
+    let (_, be) = trim_idx(all_before, 0, all_before.len());
+    let mut bb_tmp = cmp::max(be as isize - max_before_size as isize, 0) as usize;
```
Comment (sylvestre) [inline src/uu/ptx/src/ptx.rs:352]: please add a comment explaining that this block is doing

File: src/uu/ptx/src/ptx.rs [hunk 5]
```diff
@@ -316,57 +327,73 @@ fn trim_idx(s: &[char], beg: usize, end: usize) -> (usize, usize) {
...
+    tail.push_str(&tail_str);
 
     // get head
-    let max_head_size = max_after_size - after.len();
-    let (_, he) = trim_idx(&all_before_vec, 0, before_beg);
-    let mut hb_tmp = cmp::max(he as isize - max_head_size as isize, 0) as usize;
-    hb_tmp = trim_broken_word_left(&all_before_vec, hb_tmp, he);
-    let (head_beg, head_end) = trim_idx(&all_before_vec, hb_tmp, he);
-    head.push_str(&all_before[head_beg..head_end]);
+    let max_head_size = cmp::max(
```
Comment (sylvestre) [inline src/uu/ptx/src/ptx.rs:384]: same here
please add a comment explaining that this block is doing

File: src/uu/ptx/src/ptx.rs [hunk 6]
```diff
@@ -399,70 +421,94 @@ fn tex_mapper(x: char) -> String {
     }
 }
 
-fn adjust_tex_str(context: &str) -> String {
-    let ws_reg = Regex::new(r"[\t\n\v\f\r ]").unwrap();
-    let mut fix: String = ws_reg.replace_all(context, " ").trim().to_owned();
-    let mapped_chunks: Vec<String> = fix.chars().map(tex_mapper).collect();
-    fix = mapped_chunks.join("");
-    fix
+fn format_tex_field(s: &str) -> String {
```
Comment (sylvestre) [inline src/uu/ptx/src/ptx.rs:424]: please document the function

**Review by sylvestre — CHANGES_REQUESTED** [review]
Some minor nits. the code would benefit to get more comments!

It would be nice to add a few more tests too

**Comment by wishawa** [comment]
Commit 5fc007b295230730bb2b67d3db32fb7defcac6a3 removes a code that I added to try to mimic GNU ptx's behavior.
This is because I believe that the GNU version just has too many "quirks" and if we try to replicate all of them we would end up rewriting their C code in Rust line by line.
Let me know what you think.

File: src/uu/ptx/src/ptx.rs [hunk 7]
```diff
@@ -169,9 +169,9 @@ fn get_config(matches: &clap::ArgMatches) -> Config {
             .expect(err_msg)
             .to_string();
     }
-    if matches.is_present(options::IGNORE_CASE) {
+    if matches.is_present(options::FLAG_TRUNCATION) {
         config.trunc_str = matches
-            .value_of(options::IGNORE_CASE)
+            .value_of(options::FLAG_TRUNCATION)
```
Comment (sylvestre) [inline src/uu/ptx/src/ptx.rs:174]: do you think you could add a test to cover this case? thanks

**Comment by sylvestre** [comment]
Testing it, I just found this bug:
https://github.com/uutils/coreutils/issues/2049

**Comment by sylvestre** [comment]
hyperfine confirms your improvemens
```
hyperfine --warmup 2 --max-runs 10 'cp  sample.txt.orig sample.txt &&  /tmp/coreutils-old ptx -G sample.txt  > /dev/null' 'cp  sample.txt.orig sample.txt && ptx -G sample.txt > /dev/null' 'cp  sample.txt.orig sample.txt && ~/dev/debian/coreutils/target/release/coreutils ptx -G sample.txt  > /dev/null'
Benchmark #1: cp  sample.txt.orig sample.txt &&  /tmp/coreutils-old ptx -G sample.txt  > /dev/null
  Time (mean ± σ):      1.353 s ±  0.028 s    [User: 1.339 s, System: 0.010 s]
  Range (min … max):    1.302 s …  1.398 s    10 runs
 
Benchmark #2: cp  sample.txt.orig sample.txt && ptx -G sample.txt > /dev/null
  Time (mean ± σ):      29.9 ms ±   0.8 ms    [User: 28.8 ms, System: 1.1 ms]
  Range (min … max):    28.9 ms …  31.5 ms    10 runs
 
Benchmark #3: cp  sample.txt.orig sample.txt && ~/dev/debian/coreutils/target/release/coreutils ptx -G sample.txt  > /dev/null
  Time (mean ± σ):      93.6 ms ±   1.2 ms    [User: 85.3 ms, System: 8.3 ms]
  Range (min … max):    92.0 ms …  95.2 ms    10 runs
 
Summary
  'cp  sample.txt.orig sample.txt && ptx -G sample.txt > /dev/null' ran
    3.13 ± 0.09 times faster than 'cp  sample.txt.orig sample.txt && ~/dev/debian/coreutils/target/release/coreutils ptx -G sample.txt  > /dev/null'
   45.23 ± 1.50 times faster than 'cp  sample.txt.orig sample.txt &&  /tmp/coreutils-old ptx -G sample.txt  > /dev/null'
```

**Comment by sylvestre** [comment]
Well done :)
